### PR TITLE
[RLlib] Fix incorrect calculations in tf_action_dist.py Categorical.

### LINKED ISSUE #24055: [RLlib] Categorical action dist samples masked actions
### What happened + What you expected to happen

The problem is here:
https://github.com/ray-project/ray/blob/6d8d7398df4f90abd008468c5b4fb1ebfa587256/rllib/models/tf/tf_action_dist.py#L90-L91

[tf.random.categorical](https://www.tensorflow.org/api_docs/python/tf/random/categorical) takes in **log probabilities** even though the name of the input variable is 'logits'. Here, self.inputs are the logits, so masked actions (where logits[a]=0) are considered valid samples by tf. This also impacts MultiCategorical since it ultimately calls this same method.

### Versions / Dependencies

ray 1.11.0 (still present in 1.12, though)
python 3.9
tf 2.7
rhel 7.9

### Reproduction script

```
# mask last 3 actions
z = tf.constant([[0.5,0.5,0.5,0,0,0]])

# current - will sample masked actions
tf.random.categorical(z,10)

# corrected - won't sample masked actions
tf.random.categorical(tf.math.log(z),10)
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by HJasperson**
## Why are these changes needed?

Nearly every method in tf_action_dist.py Categorical produces incorrect results. I have made the following changes:

- **logp: replaced incorrect function usage** The previous implementation calls [tf.nn.sparse_softmax_cross_entropy_with_logits](https://www.tensorflow.org/api_docs/python/tf/nn/sparse_softmax_cross_entropy_with_logits), which does not produce the log probabilities. I replaced this with the actual log probability formula.
- **entropy: replaced questionable formula** I've never seen entropy formulated like this and have observed strange entropy values in my training as a result. Note that the [torch version](https://github.com/ray-project/ray/blob/master/rllib/models/torch/torch_action_dist.py) does not use this odd formula. I replaced this with the usual entropy formula. I added clip_by_value to avoid errors related to log(0) that arise due to masking logits with tf.float32.min.
- **kl: replaced questionable formula** I've also never seen KL divergence formulated like this and have observed odd KL values in my training. Again, the torch version does not use this strange formula. I replaced this with a call to [TF's built-in KL divergence function](https://www.tensorflow.org/api_docs/python/tf/keras/losses/KLDivergence), after calculating probabilities from the logits.
- **_build_sample_op: replaced incorrect function inputs** [tf.random.categorical](https://www.tensorflow.org/api_docs/python/tf/random/categorical) takes in log probabilities as input (scroll down to the bottom of the docs). TF uses "logits" and "log probabilities" interchangeably, but these are _not_ the same quantities in RLlib. Thus, we need to calculate the log probabilities from self.inputs and use those in the function call. This issue is also present in [epsilon_greedy.py](https://github.com/ray-project/ray/blob/master/rllib/utils/exploration/epsilon_greedy.py). Relevant PR: #26604

Similar issues may exist in other classes in this file since I only addressed Categorical.

## Related issue number

Replaces #24278 and closes #24055

Note: In #24278 a lack of benchmarks was mentioned and was part of why the PR went stale. If such benchmarks existed, the current implementation would fail them. Lack of benchmarks is not a good reason to keep methods that are blatantly wrong.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by HJasperson** [comment]
issue still exists in v2.0

File: rllib/models/tf/tf_action_dist.py [hunk 1]
```diff
@@ -64,32 +64,25 @@ def deterministic_sample(self) -> TensorType:
...
     def logp(self, x: TensorType) -> TensorType:
-        return -tf.nn.sparse_softmax_cross_entropy_with_logits(
-            logits=self.inputs, labels=tf.cast(x, tf.int32)
-        )
+        logp = tf.math.log(tf.nn.softmax(self.inputs))
+        return tf.gather(logp, x, batch_dims=1)
 
     @override(ActionDistribution)
     def entropy(self) -> TensorType:
-        a0 = self.inputs - tf.reduce_max(self.inputs, axis=1, keepdims=True)
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/tf/tf_action_dist.py:73]: The weird implementations you are referring to are actually modified versions of a softmax.
I've not done the original implementation, but I think that this line is supposed to add some numerical stability to the softmax. But from where I stand it looks more like it would add instabilities.

Consider _self.inputs = [n, m]_ where n and m are large negative numbers. ea0 would be close to zero and z0 would also be close to zero, with p0 being instable.

@kouroshHakha do you have some insight here?

Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/tf_action_dist.py:73]: That being said, purely mathematically speaking, the softmax is invariant to translation.

**Review by ArturNiederfahrenhorst — COMMENTED** [review]
The lines that you are removing are softmaxes with translated inputs.

File: rllib/models/tf/tf_action_dist.py (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/tf_action_dist.py:73]: I've produced [a script](https://gist.github.com/ArturNiederfahrenhorst/b721c0085457c4db845542519d8f8f84) that shows differences in results.

**Comment by HJasperson** [comment]
Thanks for the softmax insight. If you decide to keep the current kl/entropy, I recommend adding an explanatory comment and possibly changing the variable names to be more meaningful. That said, my version is more readable and your script shows negligible error, so I think you could go either way.

File: rllib/models/tf/tf_action_dist.py [hunk 2]
```diff
@@ -64,32 +64,25 @@ def deterministic_sample(self) -> TensorType:
 
     @override(ActionDistribution)
     def logp(self, x: TensorType) -> TensorType:
-        return -tf.nn.sparse_softmax_cross_entropy_with_logits(
-            logits=self.inputs, labels=tf.cast(x, tf.int32)
-        )
+        logp = tf.math.log(tf.nn.softmax(self.inputs))
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/tf/tf_action_dist.py:67]: Why don't we clip by value with tf.keras.backend.epsilon() here, too?

File: rllib/models/tf/tf_action_dist.py [hunk 3]
```diff
@@ -64,32 +64,25 @@ def deterministic_sample(self) -> TensorType:
...
-        p0 = ea0 / z0
-        return tf.reduce_sum(p0 * (a0 - tf.math.log(z0) - a1 + tf.math.log(z1)), axis=1)
+        probs = tf.nn.softmax(self.inputs)
+        probsOther = tf.nn.softmax(other.inputs)
+        return tf.keras.losses.kl_divergence(probs, probsOther)
 
     @override(TFActionDistribution)
     def _build_sample_op(self) -> TensorType:
-        return tf.squeeze(tf.random.categorical(self.inputs, 1), axis=1)
+        logp = tf.math.log(tf.nn.softmax(self.inputs))
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/tf/tf_action_dist.py:84]: Same here, why don't we clip by value with tf.keras.backend.epsilon() here, too?

File: rllib/models/tf/tf_action_dist.py (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/tf_action_dist.py:73]: I approve this this change to entropy()

File: rllib/models/tf/tf_action_dist.py [hunk 4]
```diff
@@ -64,32 +64,25 @@ def deterministic_sample(self) -> TensorType:
...
     def kl(self, other: ActionDistribution) -> TensorType:
-        a0 = self.inputs - tf.reduce_max(self.inputs, axis=1, keepdims=True)
-        a1 = other.inputs - tf.reduce_max(other.inputs, axis=1, keepdims=True)
-        ea0 = tf.exp(a0)
-        ea1 = tf.exp(a1)
-        z0 = tf.reduce_sum(ea0, axis=1, keepdims=True)
-        z1 = tf.reduce_sum(ea1, axis=1, keepdims=True)
-        p0 = ea0 / z0
-        return tf.reduce_sum(p0 * (a0 - tf.math.log(z0) - a1 + tf.math.log(z1)), axis=1)
+        probs = tf.nn.softmax(self.inputs)
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/tf/tf_action_dist.py:78]: I approve this this change to kl()

File: rllib/models/tf/tf_action_dist.py (hunk 3 above)
Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/tf_action_dist.py:84]: I approve using logprobs here though as it common practice and adds stability.

File: rllib/models/tf/tf_action_dist.py (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/tf_action_dist.py:67]: Other than that, sparse_softmax_cross_entropy_with_logits also does the job here. So I don't think a change is needed.

**Review by ArturNiederfahrenhorst — COMMENTED** [review]
- I'm yet to understand why the logp method should be changed. 
- entropy() and kl() look more readable and I don't see why we should not use the standard functions provided by the tensor backends here.
- The TF OP method I think we should do, because otherwise we miss out on the added stability here

**Comment by HJasperson** [comment]
Good catch on clip_by_value. I added the missing ones.

The current logp is odd. It should be ok assuming certain mathematical conditions are met, but I really don't know if that is guaranteed, so that aspect personally makes me uneasy. Second, this function is usually used in the context of loss functions for classification, so seeing it here is weird. Finally, the 'normal' way of calculating logp is short and easy and would eliminate any weirdness/confusion.

**Review by ArturNiederfahrenhorst — APPROVED** [review]
Approved if tests turn out green.

File: rllib/models/tf/tf_action_dist.py (hunk 1 above)
Comment (sven1977) [reply rllib/models/tf/tf_action_dist.py:73]: Awesome. Thanks for the comparison script @ArturNiederfahrenhorst !
I have no objections to merging, other than my comment below on the sampling op. I think we need to change it to make action masking still work.

File: rllib/models/tf/tf_action_dist.py [hunk 5]
```diff
@@ -64,32 +64,31 @@ def deterministic_sample(self) -> TensorType:
...
+        probsOther = tf.clip_by_value(probsOther, tf.keras.backend.epsilon(), 1.0)
+        return tf.keras.losses.kl_divergence(probs, probsOther)
 
     @override(TFActionDistribution)
     def _build_sample_op(self) -> TensorType:
-        return tf.squeeze(tf.random.categorical(self.inputs, 1), axis=1)
+        probs = tf.nn.softmax(self.inputs)
+        probs = tf.clip_by_value(probs, tf.keras.backend.epsilon(), 1.0)
+        logp = tf.math.log(probs)
+        return tf.squeeze(tf.random.categorical(logp, 1), axis=1)
```
Comment (sven1977) [inline rllib/models/tf/tf_action_dist.py:91]: I think this will break action masking and is my only concern about merging this right now.

Some actions must have zero prob sometimes, if they are masked by the env.

@HJasperson

Comment (sven1977) [reply rllib/models/tf/tf_action_dist.py:91]: Can we just use 0.0 here instead of tf.keras.backend.epsilon() for the lower bound?

Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/tf_action_dist.py:91]: But probs are not allowed to be zero anyways if we put them into tf.math.log(probs) afterwards (where they will be undefined).

Comment (sven1977) [reply rllib/models/tf/tf_action_dist.py:91]: Ah, sorry, true, log probs would fail. Just felt like tf.keras.backend.epsilon (10e-7) is still too large.

**Comment by ArturNiederfahrenhorst** [comment]
Hey @HJasperson I'd do the changes myself but git is giving a cryptic error **for this PR** that I'll not resolve.

> To github.com:ray-project/ray.git
 ! [remote rejected]       HEAD -> refs/pull/27527/head (deny updating a hidden ref)
error: failed to push some refs to 'github.com:ray-project/ray.git'

-> Can you please merge master?

We're still in the discussion on how we deal with action masking in the OP.

File: rllib/models/tf/tf_action_dist.py (hunk 5 above)
Comment (HJasperson) [reply rllib/models/tf/tf_action_dist.py:91]: tf.keras.backend.epsilon is what they use in the tf source code, so that's why I used it. I ran some tests, though, and it looks like tf,math.log should return real numbers down to about 1e-37. I can replace tf.keras.backend.epsilon with 1e-37 if y'all prefer.

Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/tf_action_dist.py:91]: I think tensorflow should accept tensor initialization with -np.inf, no?

Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/tf_action_dist.py:91]: tf.constant([-np.inf]) works on my end! So I think that would be the cleanest way to go. Sorry for the back and forth here and thanks for sticking with us!

Comment (HJasperson) [reply rllib/models/tf/tf_action_dist.py:91]: Are we still talking about `tf.clip_by_value(probs, tf.keras.backend.epsilon(), 1.0)`? The reason you need this line is because tf.math.log returns -inf if you get too close to 0. The smallest (positive) number you can input and still return a real value is 1e-37. -np.inf does not address this at all.

Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/tf_action_dist.py:91]: Absolutely. Sorry I should have been more clear.
`tf.math.log(probs)` should be -np.inf where self.inputs is zero imo.

Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/tf_action_dist.py:91]: So we'll have to "manually" set the logprobs to -np.inf in these instances.

Comment (sven1977) [reply rllib/models/tf/tf_action_dist.py:91]: Can we do `tf.float32.min` instead, then? This is the number that our e.g. EpsilonExploration uses to figure out whether it should pick an action (randomly) or not?
